### ISSUE TITLE
BL652 VSP Packet drop @ higher data Rate (2800 bytes per second)

### ISSUE THREAD

**Issue by abeldsouza** [LairdCP/BL652-Applications#14]
Dear Community ,

I would like to understand the data rate that can be supported over the Laird BL 652 VSP service. I use the BL-652 as part of my product and have been using the VSP for a while now without any issues. We started out with a data rate of 300 bytes per second. Over time we are now at a data rate of 2800 bytes per second. I now see a lot of dropped bytes. I am still debugging this and have so far not been able to firmly concluded if the issue is on the BL 652 side or the receiver side. 

Some items that have been checked are listed below

On the BL 652 side, the UART RX line is tapped out and forwarded to a windows laptop and read using serial. The data is fine. So I have confirmed that the BL 652 RX pin is indeed seeing the correct data.

I tried to receive the BLE data using PunchThrough LightBlue BLE app on android. Here is see only around 1200 to 1500 bytes of data per second. This log is generated by the Light Blue Android app.

On Windows i have used the below applications
https://github.com/sensboston/BLEConsole
https://github.com/noble/noble-device

Both these applications receive around 1000 odd bytes similar to the light blue. That's when i started to think if the issue is actually on the BL 652 side. 

While I have yet to do a lot of debugging my query is just to understand the data rates that can be supported and if there have been similar observations with the vsp.

Thanks,
Abel.

**Comment by abeldsouza** [comment]
An update

So I realize that my peripheral code is based out of the git hub example 
https://github.com/LairdCP/BL652-Applications/blob/master/%24autorun%24.devkit.sample.app.sb
This example was the one I used with the development kit and proceeded with this as the base to build my peripheral code. In this example we open the vsp using BleVspOpen and then incoming uart data is directly written to vsp using BleVSpWrite

Based on information from Laird technical person, the better peripheral code to refer is
https://github.com/LairdCP/BL652-Applications/blob/master/Applications/%24autorun%24.2mphy.throughput.peripheral.sb
Here a vsp service is created in CreateVSPService and the incoming uart data is sent using BLECHARVALUENOTIFY

The 2nd one gives better results but i dont have the real answer as to what the difference is between the 2 versions. However based on feedback from laird, will migrate my peripheral code in coming weeks to the new option to see if it helps to get the required throughput.


Currently my code works well for upto 800 bytes of data per second. I tried a few things using the current code

Set latency to 0 (at+cfg 116 0)
Set the extended data length (at+cfg 211 247 and at+cfg 212 244)
Use rts / cts

None of these made any difference. so looks like the code migration is the next thing to try.

Thanks,
Abel.

**Comment by abeldsouza** [comment]
A small update on this issue hoping that it would help someone down the line. We have fixed the issue of packet drop over BLE. some of the important findings below.

2800 bytes per second actually equates to a lot more bytes when data is sent over BLE. We generally work with a 10 byte packet structure and so every 10 bytes was actually being converted to 30+ bytes (i forget the exact number) when streamed over ble after the addition of the protocol bytes. This was checked using a packet capture on the receiver side.


We were using a baud rate of 115200. Changed now to 1000000 as per the baud rate used in the Laird tests. Link below.
https://www.lairdconnect.com/documentation/throughput-testing-bluetooth-5-le-2m-phy-feature-bl652pdf

There was one issue with the way data was sent over the vsp. the host stated to dump data immediately once the ble connection was opened. however it took an additional 3 to 4 seconds for the BLE connection to get authenticated and then to subscribe to the characteristic. By this time the vsp buffers were already being loaded with data and the data transmission started from an already overloaded  state and this error kept propagating and disrupting the communication. We now start the data stream only after the subscription to the particular characteristic. 

A few other minor improvements in the code.

After all these changes we still could not achieve the required data rate. But we saw significant improvement. As the final fix we enabled the DLE (Data length extension) in the BL652 and this now works well to stream data at the required rate. With the data size larger the overall load on chip is much lesser (related to the 10 bytes being converted to 32+ bytes)

Thanks,
Abel.
